### ISSUE TITLE
Possible issues in desktopCapturer on Windows 10

### ISSUE THREAD

**Issue by webarchymeta** [electron/electron#6599]
- Electron version: 1.3.0
- Operating system: Windows 10

When trying to capture individual window:

1) All chrome based windows are black (including opera, opera-dev, electron based apps). rectangulars.
2) Microsoft edge browser window is not shown in the list returned by calling `getSources`

Maybe my system is up for to long that something went wrong. Can someone confirm this?

**Comment by zcbenz** [comment]
Can you try whether this Chrome app has the same problem on your machine?
https://github.com/GoogleChrome/chrome-app-samples/tree/master/samples/desktop-capture

And does this happen in 1.2.x?

**Comment by webarchymeta** [comment]
I tried on the same machine after rebooting, the problem is still there for electron versions (1.2.6, 1.27, and 1.30). On a different machine with Windows 10 installed, only the app (electron based) show a black screen when captured, the screen of chrome browser is OK. It's OK maybe because it is an old version (48.x) instance.

On either machines, the screen of edge browser are not in the list screens return from `getSources`.

I did not figure out how to run the chrome sample apps (the access to google sites are limited here), could you give me a hint, if it not too much trouble?

**Comment by zcbenz** [comment]
You should be able to just install it in Chrome webstore:
https://chrome.google.com/webstore/detail/desktop-capture-sample/mhkidniocjdaiddjckopkigjmjbadfji

Can you try it with Chrome 50 and 51 on the same machine that you run Electron? Electron is using the Chromium's code for `desktopCapturer` and I want to verify whether it is a bug of Chrome.

**Comment by webarchymeta** [comment]
I can't access google sites at present (no vpn is available for a while)...

**Comment by charles-at-linknext** [comment]
Hi, 

We bumped into the same problem too.

Electron version: 1.3.3
OS: Windows 10

However, thumbnail of screen type (i.e. entire screen) was not black and looked fine. Did try Chrome 50 and 51 with the Desktop Capture Sample. 50 had black thumbnail, but 51 had fine one.

**Comment by gtk2k** [comment]
Chrome M51
this problem fixed.
https://groups.google.com/forum/#!msg/discuss-webrtc/GdsmhrVaxdU/q99shod9OAAJ

same problem on Electron version 1.3.4.

**Comment by webarchymeta** [comment]
The problems are still there after upgrading electron to v1.3.5

**Comment by gtk2k** [comment]
same prolem 1.4.0

**Comment by kevinsawicki** [comment]
@webarchymeta @gtk2k so the windows are black? Is this happening consistently or only intermittently?

**Comment by webarchymeta** [comment]
It happens consistently across all Chromium based browser views on Windows (10, at least), including Opera, Electron and, of course Chrome. Linux based systems are fine.

**Comment by gtk2k** [comment]
consistently.
I search this plobrem.

The cause of this problem is "Aura".
https://bugs.chromium.org/p/chromium/issues/detail?id=649312

**Comment by djpereira** [comment]
I can reproduce this issue with Windows 10 Version 1511.

Sharing the entire screen works, as well as sharing the command prompt, notepad, etc.

However, applications like Chrome, HipChat, Sticky Notes and Visual Studio Code are shared/displayed as just the border around the app itself with black content, both for the thumbnail and the video stream.

Additionally, applications like Outlook 2016, GitHub, Google Hangouts and Microsoft Edge are not even shown in the list of sources.

I tried using Chrome 54 and screen sharing plug-in to compare and I found out that the same apps that have issues with black content are shared correctly by Chrome, however they still have problems with the thumbnails.

Chrome 54 is also missing Outlook 2016, Edge, etc from the list.

**Comment by nuthinking** [comment]
Same issue here: Electron 1.6, Chrome 56, and Windows 10 (1607).

Recorded videos on Windows using chromeMediaSource: "desktop" are black.

When using chromeMediaSource: "screen" on the same Windows 10 machine, the videos are recorded but the content is offset by a couple of dozen pixels to the right despite the window is fullscreen. I tried the other DisplayCaptureSurfaceType but they don't provide any further option to capture the entire screen.

**Comment by alexlee85** [comment]
same problem here:  electron 1.4.14

**Comment by djpereira** [comment]
Updated to Electron 1.6.1, and now I can share Outlook 2016. However, I can't share HipChat or Chrome in Windows 10 (they are also missing the thumbnail).

I tried with a Windows 7 system I was able to share Chrome but not HipChat.

Edge still does not show up in the list.

**Comment by ondravondra** [comment]
We ran into this problem when we tried to capture the electron app itself and the result was black rectangle. We used `app.disableHardwareAcceleration()` which fixed it.

**Comment by webarchymeta** [comment]
@ondravondra I tried your method, but I am still having the same problems with other chromium based applications, except for the very app that has its hardware acceleration disabled.

Therefore, the problem seems to be related to electron's or chrome's incapability of capturing (some of?) hardware accelerated windows on Windows OS ...

**Comment by djpereira** [comment]
I tried this with Electron 1.6.6 and Windows 10 Version 1607.
I am still not able to share Chrome or other electron-based applications (e.g. Visual Studio Code).
Full screen sharing and some other applications work well.
I tried screen sharing with Chrome 58.0.3029.81 and https://hangouts.google.com/ and it works well.

**Comment by djpereira** [comment]
If this helps at all, I tried the settings below in https://github.com/electron/electron/blob/8359c72347372176e3cb27735caa19d967dae762/atom/browser/api/atom_api_desktop_capturer.cc#L60 and it did not make a difference for either the thumbnails or the actual window sharing.
```
options.set_disable_effects(true);
options.set_allow_directx_capturer(true);
```

**Comment by djpereira** [comment]
In https://github.com/electron/electron/pull/8501/commits/d4e3dade394f24b2b452931ef5abe23b11b7b47c native_desktop_media_list was updated to support the most recent changes in Chromium 56.
If you look at their version of native_desktop_media at https://cs.chromium.org/chromium/src/chrome/browser/media/webrtc/native_desktop_media_list.cc?rcl=a69c78443bceeb4444cd8083e2033c4895d94ab2&rlf=75b3b45a4797bd19d3acea459082dd4914b2b07e it seems like they have additional code to manage "aura" windows. Does anyone know if that would be needed to solve this issue?

**Comment by francescoriglietti** [comment]
i encountered the same problem. @djpereira have you been able to advance in the solution of the problems? it's still present in 1.7.3

**Comment by djpereira** [comment]
Yes, the issue is still present in 1.7.3.

**Comment by KiranNiranjan** [comment]
+1 The issue is still present in electron **1.7.8** 😞  (Windows 10)

**Comment by andrei-cacio** [comment]
does anyone have an idea about why is this happening?

**Comment by andrei-cacio** [comment]
ok so I am running a Windows 10 laptop with Intel integrated graphic chip and a nVidia dedicated graphic card. If I disable the dedicated graphic card I can capture electron apps. However I am still getting black capture for chrome windows.

**Comment by andrei-cacio** [comment]
and if I open chrome in Compatibility mode for Windows 8 I can share the chrome window

**Comment by levinunnink** [comment]
Any further updates on this. It's hitting our app.

**Comment by andrei-cacio** [comment]
@levinunnink probably a good workaround is to simply capture the entire screen for now

**Comment by arnaudbud** [comment]
another workaround is the use of the flag **--disable-gpu-compositing** in the Chrome shortcut, as well as in all the Chromium based applications (Paragon, Electron...) that could be potentially being screen shared over your Electron application. This leads us to the root cause @gtk2k mentioned... https://bugs.chromium.org/p/chromium/issues/detail?id=649312#c5 and the place to track the issue  and provide your feedback https://bugs.chromium.org/p/chromium/issues/detail?id=297173

**Comment by zengming00** [comment]
--disable-gpu-compositing not work, still black ,  on win10 Electron 1.7.11

**Comment by miaulightouch** [comment]
try `--disable-direct-composition` instead

and can we apply this with `app.commandLine.appendSwitch`? it's not working.

**Comment by vjroick** [comment]
Tried a different approach,
Your can use below code to take screenshot and show it on screen, later on you can save it by FileSaver.js

`const remote = require("electron").remote;

remote.getCurrentWindow().capturePage(function(buf) {
        // console.log(buf.toPng())
        imgElemSelector.setAttribute("src", buf.toDataURL());
})`

**Comment by icecoll** [comment]
same issue on electron 1.8.7 win 7, I find a workground [here](https://groups.google.com/forum/#!topic/discuss-webrtc/Q0SdEkAakak). Enable Aero makes it work.
[(How To Enable Window Transparency (Aero) On Windows 7)](http://www.tomshardware.com/faq/id-2066988/enable-window-transparency-aero-windows.html)

**Comment by codebytere** [comment]
Given that there exist workarounds for this, and that this isn't currently on our roadmap, i'm going to go ahead and label this a `wontfix`. However, should someone determine a workable solution we would be very open to accepting a PR!

**Comment by friksa** [comment]
There is no workable workaround for this issue.  When you offer a tool that allows for a web conference, it is not acceptable to the user to change the configuration of their system to make any feature work.  Please reconsider fixing this.

**Comment by pailjin** [comment]
> I can reproduce this issue with Windows 10 Version 1511.
> 
> Sharing the entire screen works, as well as sharing the command prompt, notepad, etc.
> 
> However, applications like Chrome, HipChat, Sticky Notes and Visual Studio Code are shared/displayed as just the border around the app itself with black content, both for the thumbnail and the video stream.
> 
> Additionally, applications like Outlook 2016, GitHub, Google Hangouts and Microsoft Edge are not even shown in the list of sources.
> 
> I tried using Chrome 54 and screen sharing plug-in to compare and I found out that the same apps that have issues with black content are shared correctly by Chrome, however they still have problems with the thumbnails.
> 
> Chrome 54 is also missing Outlook 2016, Edge, etc from the list.

what's the progress? fine or not? solution please. thanks

**Comment by pailjin** [comment]
@djpereira help.

**Comment by CapOM** [comment]
For "2. Microsoft edge browser window is not shown in the list returned by calling getSources" see: https://bugs.chromium.org/p/chromium/issues/detail?id=700037

**Comment by CapOM** [comment]
For "1. All chrome based windows are black (including opera, opera-dev, electron based apps). rectangulars." it is because on win10, the MS api `PrintWindow` (https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-printwindow) does not work. (well it works on some old apps or if an app does not use directx or GL). So the idea on Win10 to support application sharing is to use the "full screen capturer" (which is based on DirectX) and then do cropping. Of course this only work if the app window is on the foreground (not covered). If it is covered it still tries to fallback to PrintWindow which results on black window.

See https://bugs.chromium.org/p/chromium/issues/detail?id=890928#c13

In short as a test try to make the app window small, centered on the screen and not covered and then this should work. Otherwise open a bug in chromium.

**Comment by levinunnink** [comment]
It would be awesome if Electron could provide some sort of workaround for this issue. As it currently stands the desktopCapturer API is unusable when it comes to reliably capturing windows. This is a huge blocker for us as our customers need a more secure way to share their screen then displaying the entire desktop.

**Comment by CapOM** [comment]
@levinunnink you can click the `star` on http://crbug.com/890928 so it gets more attention

**Comment by levinunnink** [comment]
Any update on this issue?

**Comment by freddy-daniel** [comment]
We need a solution . It's still present in electron version **6.0.10 !**  

### It's a major issue since 2016

**Comment by peererror** [comment]
please fix this issue.

**Comment by keermanish** [comment]
Any updates on this issue. Not sure the exact scenario/steps but still some times this blank screen issue gets reproduce.

**Comment by levinunnink** [comment]
I've pretty much abandoned the desktopCapturer API and switched to embedding FFMPEG. It's a shame.

**Comment by CapOM** [comment]
It is now working:
https://bugs.chromium.org/p/webrtc/issues/detail?id=10734
https://bugs.chromium.org/p/chromium/issues/detail?id=700037

Would need to back port to electron/chromium the CLs attached to them.
